packages/vite/src/node/plugins/define.ts

```diff
@@ -37,9 +37,15 @@ export function definePlugin(config: ResolvedConfig): Plugin {
     ...importMetaKeys
   }
 
+  // sort and reverse to make a more clear replacement matches first
+  // for example:
+  // ['process.env.', 'process.env.var'] should be ordered as
+  // ['process.env.var', 'process.env.'] to make the longer key be replaced earlier
   const pattern = new RegExp(
     '\\b(' +
       Object.keys(replacements)
+        .sort()
+        .reverse()
```

Comment:
`reverse` here makes the logic harder to reason about. It should be fixed by just moving `...userDefine` and `...importMetaKeys` before `process.env` replacements.

Reply:
The user could define `process.env.` too. If I did so the user defined `process.env.` will be override.

Reply:
Isn't that what they want though? There's no good reason to do it and it will probably break, but that's what they are trying to do...

Reply:
Also I think we'd better not depend on the object keys' literal order passed by the user.

For example:
![image](https://user-images.githubusercontent.com/18677354/108220802-16f07580-7172-11eb-8f2a-e483d13edd53.png)
will cause different results.

Reply:
That's the desired behavior. Object key order matters here. Node.js/V8 respects object keys order during enumeration.

Reply:
OK